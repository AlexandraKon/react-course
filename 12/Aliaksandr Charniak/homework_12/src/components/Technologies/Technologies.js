import React, { useEffect } from 'react'
import { useDispatch } from 'react-redux'
import { Link } from 'react-router-dom'

import { getCorrectedArray } from '../../utils/utils'
import { ACTION_GET_TECHNOLOGIES_REQUESTED } from '../../redux/technologies/actions'

export const Technologies = (props) => {
  const { data } = props.data
  const { state } = props.location
  const arr = getCorrectedArray(data, 'technologies')
  const dispatch = useDispatch()
  useEffect(() => {
    dispatch(ACTION_GET_TECHNOLOGIES_REQUESTED(state?.techUrl))
  }, [dispatch, state?.techUrl])
  return (
    <div>
      {arr &&
        arr.map((item) => (
          <ul key={item.id}>
            <li>Id: {item.id}</li>
            <li>Name: {item.name}</li>
            <li>Description: {item.description}</li>
            <li>Expansion: {item.expansion}</li>
            <li>Age: {item.age}</li>
            <li>
              Develops In:
              <Link
                to={{
                  pathname: `/technologies/${item.id}/develops_in/${item.develops_in.substring(
                    item.develops_in.lastIndexOf('/') + 1
                  )}`,
                  state: {
                    structUrl: item.develops_in,
                  },
                }}
              >
                {item.develops_in.substring(item.develops_in.lastIndexOf('/') + 1)}
              </Link>
            </li>
            <li>
              Cost:
              {Object.keys(item.cost).map((key, i) => {
                return (
                  <ul key={i}>
                    <li>
                      {key}: {item.cost[key]}
                    </li>
                  </ul>
                )
              })}
            </li>
            <li>Build time: {item.build_time}</li>
            {item.applies_to && (
              <li>
                Applies To:
                {item.applies_to.map((it, i) => {
                  const unitName = it.substring(it.lastIndexOf('/') + 1)
                  return (
                    <ul key={i}>
                      <Link
                        to={{
                          pathname: `/technologies/${item.id}/applies_to/${
                            it.includes('civilization') ? 'civilization' : 'unit'
                          }/${unitName}`,
                          state: {
                            unitUrl: it,
                          },
                        }}
                      >
                        {unitName}
                      </Link>
                    </ul>
                  )
                })}
              </li>
            )}
          </ul>
        ))}
    </div>
  )
}

export default Technologies
