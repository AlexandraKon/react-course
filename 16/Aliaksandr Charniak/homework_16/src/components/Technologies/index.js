import { connect } from 'react-redux'

import Technologies from './Technologies'

const mapStateToProps = (state) => ({
  data: state.technologies,
})

export default connect(mapStateToProps)(Technologies)
