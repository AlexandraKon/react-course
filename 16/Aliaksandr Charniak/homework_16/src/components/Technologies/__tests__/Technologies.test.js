import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import configureMockStore from 'redux-mock-store'

import { Provider } from 'react-redux'

import Technologies from '..'

configure({ adapter: new Adapter() })

const mockStore = configureMockStore()
const store = mockStore({})

describe('<Technologies /> mounting behaviour', () => {
  const props = {
    data: {
      technologies: [
        {
          id: 1,
          name: 'Crossbowman',
          description: 'Upgrade to Crossbowman',
          expansion: 'Age of Kings',
          age: 'Castle',
          develops_in: 'https://age-of-empires-2-api.herokuapp.com/api/v1/structure/archery_range',
          cost: {
            Food: 125,
            Gold: 75,
          },
          build_time: 35,
          applies_to: ['https://age-of-empires-2-api.herokuapp.com/api/v1/unit/archer'],
        },
        {
          id: 2,
          name: 'Thumb Ring',
          description: 'Faster reload time (10-20%) and 100% accuracy',
          expansion: 'Age of Kings',
          age: 'Castle',
          develops_in: 'https://age-of-empires-2-api.herokuapp.com/api/v1/structure/archery_range',
          cost: {
            Food: 300,
            Wood: 250,
          },
          build_time: 45,
          applies_to: [
            'https://age-of-empires-2-api.herokuapp.com/api/v1/unit/archer',
            'https://age-of-empires-2-api.herokuapp.com/api/v1/unit/cavalry_archer',
          ],
        },
      ],
      units: 'https://age-of-empires-2-api.herokuapp.com/api/v1/units',
      structures: 'https://age-of-empires-2-api.herokuapp.com/api/v1/structures',
      civilizations: 'https://age-of-empires-2-api.herokuapp.com/api/v1/civilizations',
    },
    location: false,
  }

  let wrapper

  beforeEach(() => {
    wrapper = shallow(
      <Provider store={store}>
        <Technologies {...props} />
      </Provider>
    )
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render 1 <Technologies /> component', () => {
    expect(wrapper).toHaveLength(1)
  })
})
