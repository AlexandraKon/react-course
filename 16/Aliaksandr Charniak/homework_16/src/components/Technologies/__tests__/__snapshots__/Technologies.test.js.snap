// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Technologies /> mounting behaviour should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(Technologies)
    data={
      Object {
        "civilizations": "https://age-of-empires-2-api.herokuapp.com/api/v1/civilizations",
        "structures": "https://age-of-empires-2-api.herokuapp.com/api/v1/structures",
        "technologies": Array [
          Object {
            "age": "Castle",
            "applies_to": Array [
              "https://age-of-empires-2-api.herokuapp.com/api/v1/unit/archer",
            ],
            "build_time": 35,
            "cost": Object {
              "Food": 125,
              "Gold": 75,
            },
            "description": "Upgrade to Crossbowman",
            "develops_in": "https://age-of-empires-2-api.herokuapp.com/api/v1/structure/archery_range",
            "expansion": "Age of Kings",
            "id": 1,
            "name": "Crossbowman",
          },
          Object {
            "age": "Castle",
            "applies_to": Array [
              "https://age-of-empires-2-api.herokuapp.com/api/v1/unit/archer",
              "https://age-of-empires-2-api.herokuapp.com/api/v1/unit/cavalry_archer",
            ],
            "build_time": 45,
            "cost": Object {
              "Food": 300,
              "Wood": 250,
            },
            "description": "Faster reload time (10-20%) and 100% accuracy",
            "develops_in": "https://age-of-empires-2-api.herokuapp.com/api/v1/structure/archery_range",
            "expansion": "Age of Kings",
            "id": 2,
            "name": "Thumb Ring",
          },
        ],
        "units": "https://age-of-empires-2-api.herokuapp.com/api/v1/units",
      }
    }
    location={false}
  />
</ContextProvider>
`;
